datafusion/physical-plan/src/coalesce_batches.rs

```diff
@@ -364,90 +393,60 @@ impl BatchCoalescer {
         Arc::clone(&self.schema)
     }
 
-    /// Add a batch, returning a batch if the target batch size or limit is reached
-    fn push_batch(&mut self, batch: RecordBatch) -> Result<Option<RecordBatch>> {
-        // discard empty batches
-        if batch.num_rows() == 0 {
-            return Ok(None);
-        }
+    fn push_batch(&mut self, batch: RecordBatch) -> CoalescerState {
+        let batch = gc_string_view_batch(&batch);
```

Comment:
I think the previous semantics of checking limit before calling `gc_string_view_batch` make more sense than gc'ing the batch if the batch will be ignored

Reply:
I don't have a deep knowledge about gc'ing the batch. If I move the gc() after limit check, and if a batch leads to a limit-reached case, the lastly added batch wouldn't be handled by gc(). This isn't a problem?

Reply:
gc'ing basically compacts string data together so it requires less memory to store but requires a copy of the values

I think the implication of doing this gc before checking for limit is that we may copy the last batch worth of data prior to a limit being hit just to throw it away. However, the plan is about to shut down anyways (will switch to `CoalesceBatchesStreamState::Exhausted`) so I don't think this is a major problem